### DISCUSSION TITLE
ApiResponse 팩토리 메서드 형식 관련 제안

### DISCUSSION THREAD

**Discussion by hsjkdss228** [wanted-pre-onboarding-backend-team-s/bab-doduk#8] [General]
회원가입 API 응답 반환 로직을 작성하는 중 저희가 사용하고 있는 ApiResponse 객체에 대해 협의가 필요할 것 같은 부분이 있어 제안드립니다.

현재 `ApiResponse<T>` 인스턴스를 생성할 수 있는 팩토리 메서드는 다음의 두 종류입니다.

```java
public static <T> ApiResponse<T> toResponse(T data) {
    return new ApiResponse<T>(HttpStatus.OK.value(), data);
}

public static <T> ApiResponse<T> ok() {
    return new ApiResponse<T>(HttpStatus.NO_CONTENT.value());
}
```

회원가입 API를 작성하는 과정에서 `HttpStatus.CREATED` 상태 코드를 적용하는 팩토리 메서드를 생성해 사용하려 했으나 다음과 같은 가독성 문제가 나타났습니다.

```java
public static <T> ApiResponse<T> created(T data) {
    return new ApiResponse<T>(HttpStatus.CREATED.value(), data);
}

// ??!?
// 메서드명에서 상태 코드 200과 동일한 이름을 사용하고 있어 식별에 혼동이 있을 것 같습니다.
public static <T> ApiResponse<T> ok() {
    return new ApiResponse<T>(HttpStatus.NO_CONTENT.value());
}
```

발견된 ApiResponse 팩토리 메서드에 이슈에 대한 몇 가지 개선안을 공유드리고자 합니다. 이에 대해 의견 공유하여 주시면 감사하겠습니다!

1. ApiResponse 팩토리 메서드 명칭으로 상태 코드와 동일한 명칭 사용

```java
public static <T> ApiResponse<T> ok(T data) {
    return new ApiResponse<T>(HttpStatus.OK.value(), data);
}

public static <T> ApiResponse<T> created(T data) {
    return new ApiResponse<T>(HttpStatus.CREATED.value(), data);
}

public static <T> ApiResponse<T> noContent() {
    return new ApiResponse<T>(HttpStatus.NO_CONTENT.value());
}

// ...
```

2. data를 받거나 받지 않는 두 개의 `ApiResponse.toResponse()`를 정의하고 상태 코드를 외부에서 전달

```java
public static <T> ApiResponse<T> toResponse(HttpStatusCode statusCode) {
    return new ApiResponse<T>(statusCode.value());
}

public static <T> ApiResponse<T> toResponse(HttpStatusCode statusCode, T data) {
    return new ApiResponse<T>(statusCode.value(), data);
}
```

**Comment by buhee1029** [comment]
저의 생각은 1번 같은 경우는 메서드 만으로 어떤 응답상태인지 확인할 수 있다는 장점이 있지만 또 다른 응답상태과 동일한 상황이 나올 경우마다 메서드를 추가해야하는 단점이 있을 것 같습니다. 2번 방법으로 각 결과에 대한 상태코드를 전달하는 것이 좋은 방법이라고 생각됩니다.

**Comment by newnyee** [comment]
저도 부희님과 동일한 의견입니다!
2번의 코드 사용시 파라미터의 HttpStatus만으로도 어떤 응답상태를 반환해줄것인지 예측할 수 있고, 메소드들이 상태코드에 종속되지 않아서 2번이 더 좋은 방법인것같아요!

**Comment by kmss6905** [comment]
저는 1번을 선호합니다!
이유는 1번은 사용자가 상태코드를 직접 정의해서 주입할 수 없기 때문에 저희가 정한 상태코드와 바디만을 사용할 수 있습니다. 그렇기 때문에 사용자가 2번과 같이 마음대로 상태코드를 정의해서 넣는 상황을 막을 수 있다고 생각합니다 😄

**Comment by hsjkdss228** [comment]
남겨주신 의견들과 작업 공유 시간에 다같이 나눈 의견들을 종합적으로 판단했을 때, 1번 방식을 적용해 반영했습니다. 2번의 경우 상태 코드의 결정이 메서드 외부에 의존적으로 이루어지기 때문에, HttpStatus.NO_CONTENT 응답 코드와 함께 응답 본문을 제공하는 것과 같이 잘못된 응답 코드를 전달하는 경우에 대응하기 어려울 수 있다고 판단했습니다.
